### PR TITLE
Texture and Tiling support in SpriteRendererComponent along with Seri…

### PR DESCRIPTION
**Pull request by MohitSethi99**
…alization and Deserialization of Scene.

- Also added ImGui::ImageButton in SpriteRendererComponent UI which displays currently applied texture just like Unreal Engine.

### CODE REVIEW

File: Hazel/src/Hazel/Scene/Scene.cpp [hunk 1]
```diff
@@ -94,7 +94,7 @@ namespace Hazel {
 		{
 			auto [transform, sprite] = group.get<TransformComponent, SpriteRendererComponent>(entity);
 
-			Renderer2D::DrawQuad(transform.GetTransform(), sprite.Color);
+			Renderer2D::DrawQuad(transform.GetTransform(), sprite.Texture, sprite.TilingFactor, sprite.Color);
```
Comment (FahimFuad) [inline Hazel/src/Hazel/Scene/Scene.cpp:97]: I think you have to update this, as there is mouse-picking in Hazel. The entity ID is missing here 🙂

File: Hazelnut/src/Panels/SceneHierarchyPanel.cpp [hunk 2]
```diff
@@ -323,6 +339,41 @@ namespace Hazel {
...
+				ImGui::PushStyleVar(ImGuiStyleVar_ItemSpacing, ImVec2{ 0, 0 });
+
+				ImGui::Text("Texture");
+				const ImVec2 buttonSize = { 80, 80 };
+				ImGui::SameLine(ImGui::GetWindowWidth() * 0.6f);
+
+				ImGui::PushStyleColor(ImGuiCol_Button, ImVec4{ 0.25f, 0.25f, 0.25f, 1.0f });
+				ImGui::PushStyleColor(ImGuiCol_ButtonHovered, ImVec4{ 0.35f, 0.35f, 0.35f, 1.0f });
+				ImGui::PushStyleColor(ImGuiCol_ButtonActive, ImVec4{ 0.25f, 0.25f, 0.25f, 1.0f });
+				if(ImGui::ImageButton((ImTextureID)id, buttonSize, { 0, 0 }, { 1, 1}, 0))
```
Comment (FahimFuad) [inline Hazelnut/src/Panels/SceneHierarchyPanel.cpp:356]: This (line - 356) is wrong. The image is inverted in this method. This should be ⏬ 
`if (ImGui::ImageButton((ImTextureID)id, buttonSize, { 0, 1 }, { 1, 0 }, 0, {1, 0, 1, 1}))`

**Review by FahimFuad — CHANGES_REQUESTED** [review]

File: Hazel/src/Hazel/Scene/SceneSerializer.cpp [hunk 3]
```diff
@@ -239,6 +241,15 @@ namespace Hazel {
 				{
 					auto& src = deserializedEntity.AddComponent<SpriteRendererComponent>();
 					src.Color = spriteRendererComponent["Color"].as<glm::vec4>();
+					auto textureFilePath = spriteRendererComponent["TextureFilepath"];
+					if(textureFilePath)
+					{
+						std::string textureFilepath = textureFilePath.as<std::string>();
+						src.SetTexture(textureFilepath);
+					}
```
Comment (MeAndMyUnicorn) [inline Hazel/src/Hazel/Scene/SceneSerializer.cpp:249]: My proposition:

```cpp
auto textureFilePath = spriteRendererComponent["TextureFilepath"].as<std::string>();
if (!YAML::IsNullString(textureFilePath))
    src.SetTexture(textureFilePath);
```

File: Hazelnut/src/Panels/SceneHierarchyPanel.cpp [hunk 4]
```diff
@@ -323,6 +339,41 @@ namespace Hazel {
...
+
+			// Texture
+			{
+				const uint32_t id = component.Texture == nullptr ? 0 : component.Texture->GetRendererID();
+
+				ImGui::PushStyleVar(ImGuiStyleVar_ItemSpacing, ImVec2{ 0, 0 });
+
+				ImGui::Text("Texture");
+				const ImVec2 buttonSize = { 80, 80 };
+				ImGui::SameLine(ImGui::GetWindowWidth() * 0.6f);
```
Comment (MeAndMyUnicorn) [inline Hazelnut/src/Panels/SceneHierarchyPanel.cpp:351]: "ImGui::GetWindowWidth()" is deprecated. Use the following example instead:

```cpp
ImGui::SameLine(ImGui::GetContentRegionAvail().x * 0.67f);
```

**Review by MeAndMyUnicorn — CHANGES_REQUESTED** [review]
There is one thing or another that could be done differently.

File: Hazel/src/Hazel/Scene/Components.h [hunk 5]
```diff
@@ -45,11 +46,22 @@ namespace Hazel {
...
 		SpriteRendererComponent(const glm::vec4& color)
 			: Color(color) {}
+		
+		void SetTexture(std::string& filepath)
+		{
+			HZ_CORE_WARN("{0}", filepath);
+			Texture = Texture2D::Create(filepath);
+			TextureFilepath = filepath;
+		}
+		void RemoveTexture() { Texture = nullptr; }
```
Comment (MeAndMyUnicorn) [inline Hazel/src/Hazel/Scene/Components.h:64]: You should also reset the other values.

```cpp
void RemoveTexture()
{
    Texture = NULL;
    TextureFilepath.clear();
    TilingFactor = 1.0f;
}
```

**Review by MeAndMyUnicorn — CHANGES_REQUESTED** [review]

File: Hazel/src/Hazel/Scene/Components.h [hunk 6]
```diff
@@ -45,11 +46,22 @@ namespace Hazel {
 	struct SpriteRendererComponent
 	{
 		glm::vec4 Color{ 1.0f, 1.0f, 1.0f, 1.0f };
+		Ref<Texture2D> Texture = nullptr;
```
Comment (MeAndMyUnicorn) [inline Hazel/src/Hazel/Scene/Components.h:49]: Why is a "nullptr" used here?!

"NULL" also works fine.

Comment (FahimFuad) [reply Hazel/src/Hazel/Scene/Components.h:49]: `nullptr` is fine in my opinion. 🙂

**Comment by TheCherno** [comment]
Obsolete.

**Pull request closed without merging**